### ISSUE TITLE
Please document how to write unit tests for Bolt APIs.

### ISSUE THREAD

**Issue by offbyone** [slackapi/bolt-python#380]
I am looking to write unit tests for my Bolt API, which consists of an app with several handlers for events / messages. The handlers are both _decorated_ using the `app.event` decorator, _and_ make use of the `app` object to access things like the `db` connection that has been put on it. For example:

```python
# in main.py
app = App(
    token=APP_TOKEN,
    signing_secret=SIGNING_SECRET,
    process_before_response=True,
    token_verification_enabled=token_verification_enabled,
)
app.db = db

# in api.py:
from .main import app

@app.command("/slashcommand")
def slash_command(ack, respond):
    ack()
    results = app.db.do_query()
    respond(...)
```

The thing is, I cannot find any framework pointers, or documentation, on how I would write reasonable unit tests for this. Presumably ones where the kwargs like `ack` and `respond` are replaced by test doubles. How do I do this?
### The page URLs

* https://slack.dev/bolt-python/

## Requirements

Please read the [Contributing guidelines](https://github.com/slackapi/bolt-python/blob/main/.github/contributing.md) and [Code of Conduct](https://slackhq.github.io/code-of-conduct) before creating this issue or pull request. By submitting, you are agreeing to those rules.

**Comment by seratch** [comment]
Hi @offbyone, thanks for taking the time to write in!

As we have been receiving feedback about testing, it's on our radar. I cannot tell when we can release built-in test support but I myself will be working on it in the latter half of this year.

At this moment, the only thing I can suggest is to do something similar with the tests this project has. You can find so many tests under this package: https://github.com/slackapi/bolt-python/tree/v1.6.1/tests/scenario_tests

The key points are:

* Have your own mock server like [this](https://github.com/slackapi/bolt-python/blob/v1.6.1/tests/mock_web_api_server.py),  which responds to Web API calls / response_url calls from test code. You don't need to use the standard HTTP server module for it.
* `App` instances in tests need to have `client` argument in its constructor. The `client`'s `base_url` needs to point the mock server's URL [this way](https://github.com/slackapi/bolt-python/blob/v1.6.1/tests/scenario_tests/test_block_actions.py#L23-L26)
* Prepare request payload example data like [this](https://github.com/slackapi/bolt-python/blob/v1.6.1/tests/scenario_tests/test_block_actions.py#L154-L189)

In addition to above,

>I cannot find any framework pointers

You can use any testing frameworks (my recommendation is pytest). The forthcoming testing support won't require your tests to rely on a specific test framework. We may provde some additional utility for pytest etc. for convenience, though.

>Presumably ones where the kwargs like ack and respond are replaced by test doubles. How do I do this?

For `respond`, you can have a local mock HTTP server for handling the request. Regarding the `ack` call, it will be eventually converted to either an HTTP response or WebSocket message (Socket Mode) but before that, you can verify the value by checking `response: Response` returned by `App#dispatch` method as [this test does](https://github.com/slackapi/bolt-python/blob/v1.6.1/tests/scenario_tests/test_block_suggestion.py#L75-L77).

I know that this approach is not so easy but it should be reasonable enough. Consider going with it for now. I hope this was helpful to you. If everything is clear for now, would you mind closing this issue? For the testing support task, I will come up with a bit more detailed issue soon.

**Comment by offbyone** [comment]
Yeah, it ... helps. I guess I'd consider, if I were you, leaving this issue open to track that there is a gap, and resolving it when that documentation exists. If you don't track issues that way, please feel free to close it, but as far as I'm concerned this is still a bug until I can go to your documentation site and see instructions on how to test my bot :D

**Comment by seratch** [comment]
>if I were you, leaving this issue open to track that there is a gap, and resolving it when that documentation exists

Yeah, I understand this way. It's also fine to keep this open for now. I have a link to this in the new issue that organizes the tasks/discussions around testing support. Thanks for the nice suggestion.

**Comment by offbyone** [comment]
I have a further question: Since this kind of testing requires constructing the `app` object differently, but the object is created at the module level, does someone there have a working pattern for how to do this in the bot code itself? Your example tests all create the app _in_ the tests, but in a real Bolt application by the time you've _imported_ the code, the App already exists and has been configured. So it's really tricky trying to sort out the right method here.

**Comment by carlos-alberto** [comment]
Hey @seratch is there any update on when this will be released?

**Comment by seratch** [comment]
@carlos-alberto I did a quick prototyping late last year. After that, we haven't had good progress due to other priorities. But we're aiming to provide a better solution sometime this year!

**Comment by I-Dont-Remember** [comment]
If it helps those who stumble on this in the future, this is an [open-source Bolt app](https://github.com/happybara-io/WorkflowBuddy/tree/main/tests) with a moderately sized test suite as an example to dig through.

We've taken the approach of keeping our core functionality in a separate module that doesn't depend on `slack_bolt`, letting us easily run tests without dealing with bolt trying to start a server or anything like that.

Then in [app.py](https://github.com/happybara-io/WorkflowBuddy/blob/b1a12f0e6847f3bf14dde61a0d22a99240dbb304/app.py#L23) you can just import the module and use the Slack @decorated functions essentially as wrappers over your unit-testable business logic.

⚠ _This only helps for unit tests, unfortunately haven't delved into integration tests, so unfortunately no review/feedback on seratch's suggestion._

**Comment by jimenezj8** [comment]
Any progress on this since the last update?

**Comment by seratch** [comment]
Hey @jimenezj8, this is still on my radar and actually I did quick prototyping last year. However, I cannot tell when we can deliver a stable solution due to our bandwidth and priorities. We will update you when we come up with something in the future.

**Comment by jimenezj8** [comment]
Thanks for the update @seratch :slightly_smiling_face: will keep my eyes peeled for another

**Comment by fr12k** [comment]
Hi i was looking into how to write unit tests for out slack bot and find this wonderful library.
https://github.com/gabrielfalcao/HTTPretty
That make's it easy to mock http requests.

Just be aware that the 1.1 version is broken if you want to use it us 1.0.5.
https://github.com/gabrielfalcao/HTTPretty/issues/425

**Comment by ChadDa3mon** [comment]
Came across this as well and figured I'd try to contribute a bit. Here's some modified code from my current app, hopefully it can help someone else. Normally I'd have all of the payload stuff in my conftest.py file to keep things clean, but for the purposes of this I figured I'd put it all in one place. 

```python
import pytest
from unittest.mock import AsyncMock
from my.production.code import easter_egg

# This is the shortcut we expect to receive from Slack (well, a heavily truncated one)
mock_shortcut = {
        "message": {
            "text": "Hello, World!",
            "ts": "1234567890.123456"
        },
        "user": {
            "id": "U123456",
            "name": "John Doe"
        },
        "channel": {
            "name": "general"
        },
        "trigger_id": "trigger_12345"
    }

# Upon receiving this shortcut, we will send a 'view' to our user
# This is what we expect it to look like
expected_modal = {
    "type": "modal",
    "callback_id": "hello-modal",
    "title": {
        "type": "plain_text",
        "text": "Greetings!"
    },
    "submit": {
        "type": "plain_text",
        "text": "Good Bye"
    },
    "blocks":[
        {
          "type": "section",
          "text": {
            "type": "plain_text",
            "text": "Frosty the Snowman",
            "emoji": True
          }
        },
        {
          "type": "video",
          "title": {
            "type": "plain_text",
            "text": "Frosty the Snowman",
            "emoji": True
          },
          "title_url": "https://www.youtube.com/watch?v=RRxQQxiM7AA",
          "description": {
            "type": "plain_text",
            "text": "Frosty the Snowman",
            "emoji": True
          },
          "video_url": "https://www.youtube.com/embed/vjscH2WBWjw?feature=oembed&autoplay=1",
          "alt_text": "Frosty the Snowman",
          "thumbnail_url": "https://i.ytimg.com/vi/bSzBBK8gC6c/hqdefault.jpg",
          "author_name": "",
          "provider_name": "YouTube",
          "provider_icon_url": "https://a.slack-edge.com/80588/img/unfurl_icons/youtube.png"
        }
    ]
}

@pytest.mark.asyncio
async def test_easter_egg():
    ack = AsyncMock()
    client = AsyncMock()
    client.views_open = AsyncMock()

    # Test my actual production function 'easter_egg'
    # This should ack the shortcut and then open a view to the user
    await easter_egg(ack, mock_shortcut, client)
    ack.assert_awaited_once # Make sure it was acknowledged
    client.views_open.assert_awaited_once_with( # Make sure the view we get back is correct
        trigger_id=mock_shortcut["trigger_id"],
        view=expected_modal
    )
```

**Comment by leon-adarcewicz** [comment]
Any progress on this since the last update?

**Comment by dsoprea** [comment]
@seratch Any feedback on the current state of unit-testing support?

**Comment by seratch** [comment]
No updates from me lately. To align with our platform roadmap and priorities, I am not actively working on the prototype right now.

**Comment by NathanSmeltzer** [comment]
This may help with testing, but I haven't tried it yet:
https://github.com/ygalblum/slack-server-mock

**Comment by diegojromerolopez** [comment]
I have been working with this slack-bolt for some months now and I also encountered the problem of no easy way to test events (in my case message events).

So, in case somebody still wants to use unittest to implement simple test cases for their Slack bots, I have created a repository (https://github.com/diegojromerolopez/slack-bolt-testcase) with some base TestCase classes that can be used with unittest standard package. Or if you don't want to install the package, I hope it inspires you about a simple way of creating tests for your events.

Of course, all feedback is welcomed.

Thank you.

**Comment by markkenangray-ui** [comment]
I built a testing utility that solves the problem described here. It's called [botlint-slack](https://www.npmjs.com/package/botlint-slack) — currently Node only but Python is on the roadmap. Includes offline Block Kit validation and a mock client for unit testing handlers without hitting the real Slack API. Zero dependencies. Would love feedback from anyone who gives it a try.

**Comment by markkenangray-ui** [comment]
Published a full guide on this if useful: https://dev.to/markturk87/the-definitive-guide-to-testing-slack-bots-and-why-its-been-so-hard-5feh
